sycl/include/CL/sycl/detail/type_traits.hpp

```diff
@@ -8,66 +8,256 @@
-using remove_const_t = typename std::remove_const<T>::type;
+struct make_unsigned_impl<
+    T, enable_if_t<is_contained<T, gtl::scalar_integer_list>::value, T>> {
+  using type = typename std::make_unsigned<T>::type;
+};
 
-template <typename T> using add_pointer_t = typename std::add_pointer<T>::type;
+template <typename T>
+struct make_unsigned_impl<
+    T, enable_if_t<is_contained<T, gtl::vector_integer_list>::value, T>> {
+  using base_type = make_unsigned_impl_t<data_type_base_t<T>>;
+  using type = change_base_t<T, base_type>;
+};
 
+// TODO Delete this specialization after solving the problems in the test
+// infrastructure.
 template <typename T>
-using iterator_category_t = typename std::iterator_traits<T>::iterator_category;
+struct make_unsigned_impl<
+    T, enable_if_t<!is_contained<T, gtl::integer_list>::value, T>> {
+  using type = T;
+};
+
+template <typename T> struct make_unsigned {
+  using new_type_wo_cv_qualifiers = make_unsigned_impl_t<remove_cv_t<T>>;
+  using type = copy_cv_qualifiers_t<T, new_type_wo_cv_qualifiers>;
+};
+
+template <typename T> using make_unsigned_t = typename make_unsigned<T>::type;
 
+// Checks that sizeof base type of T equial N and T satisfies S<T>::value
+template <typename T, int N, template <typename> class S>
+using is_gen_based_on_type_sizeof =
+    bool_constant<S<T>::value && (sizeof(data_type_base_t<T>) == N)>;
+
+// is_integral
+template <typename T>
+struct is_integral : std::is_integral<data_type_base_t<T>> {};
+
+// is_floating_point
 template <typename T>
-using iterator_value_type_t = typename std::iterator_traits<T>::value_type;
+struct is_floating_point_impl : std::is_floating_point<T> {};
+
+template <> struct is_floating_point_impl<half> : std::true_type {};
+
+template <typename T>
+struct is_floating_point
+    : is_floating_point_impl<remove_cv_t<data_type_base_t<T>>> {};
+
+// is_arithmetic
+template <typename T>
+struct is_arithmetic
+    : bool_constant<is_integral<T>::value || is_floating_point<T>::value> {};
+
+// is_pointer
+template <typename T, template <typename> class TypeSelector,
+          typename SpaceList>
+struct is_pointer_impl : std::false_type {};
+
+template <typename T, template <typename> class TypeSelector,
+          typename SpaceList>
+struct is_pointer_impl<T *, TypeSelector, SpaceList>
+    : conditional_t<TypeSelector<T>::value, std::true_type, std::false_type> {};
+
+template <typename T, access::address_space Space,
+          template <typename> class TypeSelector, typename SpaceList>
+struct is_pointer_impl<multi_ptr<T, Space>, TypeSelector, SpaceList>
+    : conditional_t<(TypeSelector<T>::value &&
+                     is_one_of_spaces<Space, SpaceList>::value),
+                    std::true_type, std::false_type> {};
+
+template <typename T, template <typename> class TypeSelector = is_arithmetic,
+          typename SpaceList = gntl::all_address_space_list>
+struct is_pointer : is_pointer_impl<remove_cv_t<T>, TypeSelector, SpaceList> {};
```

Comment:
This struct does far more than just checking whether a given type is a pointer or not. I would suggest either splitting this functionality or renaming the struct.

Reply:
Done.

Reply:
Done.